File: test/support/qr_decode.py
#!/usr/bin/env python3
"""Decode a QR code with OpenCV on behalf of the Elixir test suite.

Reads a single JSON argument describing the image to decode and prints a JSON
result on stdout. Two input shapes are supported:

  {"kind": "raw",  "path": "...", "width": W, "height": H}
  {"kind": "file", "path": "..."}

A raw input is a headerless RGB8 buffer, which lets Elixir hand pixels straight
to OpenCV without depending on an image encoder or an SVG rasteriser.

Optional keys:

  "margin": true   also run a degradation ladder (downscale + blur + contrast
                   reduction) and report how far the code survives. This is the
                   difference between "decodes in a pristine screenshot" and
                   "decodes from a phone camera pointed at a stream".
"""

import json
import sys

import cv2
import numpy as np


def load(spec):
    if spec["kind"] == "raw":
        buf = np.fromfile(spec["path"], dtype=np.uint8)
        expected = spec["height"] * spec["width"] * 3
        if buf.size != expected:
            raise ValueError(f"expected {expected} bytes, got {buf.size}")
        rgb = buf.reshape((spec["height"], spec["width"], 3))
        return cv2.cvtColor(rgb, cv2.COLOR_RGB2BGR)

    img = cv2.imread(spec["path"], cv2.IMREAD_COLOR)
    if img is None:
        raise ValueError(f"could not read image at {spec['path']}")
    return img


def decode(img):
    value, _points, _straight = cv2.QRCodeDetector().detectAndDecode(img)
    return value or None


# Smallest rendered width (px) we probe down to. A QR that still decodes after
# being squeezed this small has room to spare at full size.
DOWNSCALE_LADDER = [320, 260, 200, 160, 130, 110, 90, 70]
BLUR_LADDER = [0, 3, 5, 7, 9, 11, 13]
# Multiplicative contrast reduction toward mid grey: 1.0 is untouched.
CONTRAST_LADDER = [1.0, 0.7, 0.5, 0.35, 0.25, 0.15, 0.1]


def smallest_decodable(img, expected):
    """Smallest width still decodable *without a gap* on the way down.

    Walking the ladder and keeping any width that happened to work overstates
    the margin badly: a render that fails at full size can start decoding again
    once downscaling averages the decoration away. Only an unbroken run from
    the top is evidence.
    """
    best = None
    for width in DOWNSCALE_LADDER:
        scaled = cv2.resize(img, (width, width), interpolation=cv2.INTER_AREA)
        if decode(scaled) != expected:
            break
        best = width
    return best


def blurriest_decodable(img, expected):
    best = None
    for ksize in BLUR_LADDER:
        blurred = img if ksize == 0 else cv2.GaussianBlur(img, (ksize, ksize), 0)
        if decode(blurred) == expected:
            best = ksize
    return best


def lowest_contrast_decodable(img, expected):
    best = None
    for factor in CONTRAST_LADDER:
        faded = cv2.addWeighted(
            img, factor, np.full_like(img, 128), 1 - factor, 0
        )
        if decode(faded) == expected:
            best = factor
    return best


def main():
    spec = json.loads(sys.argv[1])
    img = load(spec)
    value = decode(img)

    result = {"value": value}

    if spec.get("margin") and value:
        result["margin"] = {
            "smallest_px": smallest_decodable(img, value),
            "max_blur_kernel": blurriest_decodable(img, value),
            "min_contrast": lowest_contrast_decodable(img, value),
        }

    print(json.dumps(result))


if __name__ == "__main__":
    try:
        main()
    except Exception as error:  # noqa: BLE001 - surfaced to Elixir as JSON
        print(json.dumps({"error": str(error)}))
        sys.exit(1)

File: test/support/qr_decode.ex
defmodule Notable.QrDecode do
  @moduledoc """
  Test helper that decodes a rendered QR code with OpenCV's `QRCodeDetector`.

  The `/qr` page is displayed live on stream for viewers to scan with real
  phones, so "it looks like a QR code" is not evidence. Every visual change to
  the QR has to be checked against an actual decoder, and that check has to be
  cheap enough to run on each change rather than once at the end.

  OpenCV lives in Python, so this module shells out to `qr_decode.py`. Pixels
  are handed over as a headerless RGB8 buffer, which keeps the Elixir side free
  of any image-encoder or SVG-rasteriser dependency.
  """

  @script Path.join(__DIR__, "qr_decode.py")

  @typedoc "A raw RGB8 image as produced by `Notable.Qr.render_rgb/2`."
  @type image :: %{width: pos_integer(), height: pos_integer(), data: binary()}

  @typedoc """
  How much abuse the code survives, from `decode/2` with `margin: true`.

  * `:smallest_px` – smallest square render (px) that still decoded
  * `:max_blur_kernel` – largest Gaussian blur kernel that still decoded
  * `:min_contrast` – lowest contrast factor (1.0 = untouched) that still decoded
  """
  @type margin :: %{
          smallest_px: pos_integer() | nil,
          max_blur_kernel: non_neg_integer() | nil,
          min_contrast: float() | nil
        }

  @doc """
  Returns true when OpenCV is importable, so tests can explain a skip rather
  than fail confusingly on a machine without it.
  """
  @spec available?() :: boolean()
  def available? do
    case System.cmd("python3", ["-c", "import cv2"], stderr_to_stdout: true) do
      {_output, 0} -> true
      _ -> false
    end
  rescue
    ErlangError -> false
  end

  @doc """
  Decodes a raw RGB8 image and returns `{:ok, decoded}` or `{:error, reason}`.

  With `margin: true` the result is `{:ok, decoded, margin}`, where the margin
  comes from a degradation ladder rather than a single pristine read.
  """
  @spec decode(image(), keyword()) ::
          {:ok, String.t()} | {:ok, String.t(), margin()} | {:error, term()}
  def decode(%{width: width, height: height, data: data}, opts \\ []) do
    path = Path.join(System.tmp_dir!(), "notable-qr-#{System.unique_integer([:positive])}.rgb")
    File.write!(path, data)

    try do
      run(%{kind: "raw", path: path, width: width, height: height}, opts)
    after
      File.rm(path)
    end
  end

  @doc """
  Decodes an image file (PNG, JPEG, …) from disk. Used for browser screenshots.
  """
  @spec decode_file(Path.t(), keyword()) ::
          {:ok, String.t()} | {:ok, String.t(), margin()} | {:error, term()}
  def decode_file(path, opts \\ []) do
    run(%{kind: "file", path: path}, opts)
  end

  defp run(spec, opts) do
    spec = Map.put(spec, :margin, Keyword.get(opts, :margin, false))

    case System.cmd("python3", [@script, JSON.encode!(spec)], stderr_to_stdout: true) do
      {output, 0} -> parse(output, spec.margin)
      {output, status} -> {:error, {:decoder_failed, status, String.trim(output)}}
    end
  end

  defp parse(output, want_margin?) do
    case JSON.decode(String.trim(output)) do
      {:ok, %{"value" => nil}} ->
        {:error, :not_decodable}

      {:ok, %{"value" => value, "margin" => margin}} when want_margin? ->
        {:ok, value, atomize(margin)}

      {:ok, %{"value" => value}} ->
        {:ok, value}

      {:ok, %{"error" => reason}} ->
        {:error, reason}

      _ ->
        {:error, {:unexpected_decoder_output, output}}
    end
  end

  defp atomize(margin) do
    %{
      smallest_px: margin["smallest_px"],
      max_blur_kernel: margin["max_blur_kernel"],
      min_contrast: margin["min_contrast"]
    }
  end
end
